lib/kino/js.ex

```diff
@@ -0,0 +1,362 @@
+        {true, false} ->
+          inline_assets
+
+        {false, true} ->
+          if loaded_assets == [] do
+            IO.warn(
+              "assets directory specified for #{inspect(env.module)}, but no files" <>
+                " found in #{inspect(assets_path)}",
+              Macro.Env.stacktrace(env)
+            )
+          end
+
+          loaded_assets
+
+        {true, true} ->
+          IO.warn(
+            "ignoring files in #{inspect(assets_path)} because #{inspect(env.module)}" <>
+              " already defines inline assets with the assets/2 macro",
+            Macro.Env.stacktrace(env)
+          )
+
+          inline_assets
+
+        {false, false} ->
+          message = ~s'''
+          no assets defined for #{inspect(env.module)}.
+
+          Make sure to either explicitly specify assets directory:
+
+              use Kino.JS, assets_path: Path.expand("path/to/assets", __DIR__)
+
+          Or define assets inline:
+
+              asset "main.js" do
+                """
+                export function init(ctx, data) {
+                  ...
+                }
+                """
+              end
+          '''
+
+          IO.warn(message, Macro.Env.stacktrace(env))
+
+          []
+      end
+
+    filenames = Enum.map(assets, &elem(&1, 0))
+    duplicates = Enum.uniq(filenames -- Enum.uniq(filenames))
+
+    if duplicates != [] do
+      duplicates = duplicates |> Enum.map(&inspect/1) |> Enum.join(", ")
+
+      IO.warn(
+        "found duplicate assets in #{inspect(env.module)}: #{duplicates}",
+        Macro.Env.stacktrace(env)
+      )
+    end
+
+    if assets != [] and "main.js" not in filenames do
+      IO.warn(
+        ~s'missing required asset "main.js" in #{inspect(env.module)}',
+        Macro.Env.stacktrace(env)
+      )
+    end
+
+    assets = Enum.uniq_by(assets, &elem(&1, 0))
+
+    dir = dir_for_module(env.module)
+
+    File.rm_rf!(dir)
+    File.mkdir_p!(dir)
+
+    {archive_path, hash} = package_assets!(dir, assets)
+
+    quote do
+      def __js_info__() do
+        %{
+          assets: %{
+            archive_path: unquote(archive_path),
+            js_path: "main.js",
+            hash: unquote(hash)
+          }
+        }
+      end
+    end
+  end
+
+  def __paths__(nil), do: []
+
+  def __paths__(path) do
+    Path.join(path, "**")
+    |> Path.wildcard()
+    |> Enum.reject(&File.dir?/1)
+    |> Enum.map(&String.replace_leading(&1, path <> "/", ""))
+    |> Enum.sort()
+  end
+
+  defp package_assets!(dir_path, assets) do
+    hash = assets_hash(assets)
+    archive_path = Path.join(dir_path, hash <> ".tar.gz")
+
+    archive_content =
+      for {filename, content} <- assets, do: {String.to_charlist(filename), content}
+
+    :ok = :erl_tar.create(archive_path, archive_content, [:compressed])
+
+    {archive_path, hash}
+  end
+
+  defp assets_hash(assets) do
+    md5_hash =
+      assets
+      |> Enum.sort_by(fn {filename, _content} -> filename end)
```

Comment:
```suggestion
      |> Enum.sort()
```

Given filenames can't be duplicate (and if they are, you want to order them somehow anyway).